### ISSUE TITLE
Not working properly in monorepo 

### ISSUE THREAD

**Issue by yukinami** [vitest-dev/vscode#110]
**Describe the bug**
Not working properly in monorepo projects. By not working propertyly means vite could not resolve path alias becasue vitest was executed in the wrong user dir which is the root dir (which should be the package/xxx dir where vite.config exists ).

**To Reproduce**
Run test in monorepo with `resolve.alias` defined in `vite.config` of sub package.

**Expected behavior**
Working properly in monorepo.

**Screenshots**
![image](https://user-images.githubusercontent.com/3241464/209111264-bacb0d9a-056d-41b4-9a1e-00c308f75180.png)
![image](https://user-images.githubusercontent.com/3241464/209112620-98258c1e-048e-490a-b2fd-4c02a6c5c63d.png)

**Environment**

(Paste info.txt content generated by [the example project](https://github.com/zxch3n/vitest-ext-basic-example))

- OS: macOS 12.0
- VSCode version: 1.74.2
- Vitest version: 0.26.0
- Vitest plugin version: 0.2.34

**Additional context**

**Comment by adrienbaron** [comment]
Hi! I've had the same issue and I fixed it by creating [a VSCode Workspace file](https://code.visualstudio.com/docs/editor/workspaces). Once you have one the extension seems to pick up it needs to use the workspace root your test is in 👍

**Comment by yukinami** [comment]
> Hi! I've had the same issue and I fixed it by creating [a VSCode Workspace file](https://code.visualstudio.com/docs/editor/workspaces). Once you have one the extension seems to pick up it needs to use the workspace root your test is in 👍

You mean using [Multi-root workspaces](https://code.visualstudio.com/docs/editor/workspaces#_multiroot-workspaces)? Can you show me an example?

**Comment by adrienbaron** [comment]
> > Hi! I've had the same issue and I fixed it by creating [a VSCode Workspace file](https://code.visualstudio.com/docs/editor/workspaces). Once you have one the extension seems to pick up it needs to use the workspace root your test is in 👍
> 
> 
> 
> You mean using [Multi-root workspaces](https://code.visualstudio.com/docs/editor/workspaces#_multiroot-workspaces)? Can you show me an example?

The project I've used that in is closed source (it's at my workplace), but it's literally just creating the file like indicated in the doc and setting one of the root path to be the folder where Vitest is installed in your repo 😉

**Comment by dohooo** [comment]
Same here

**Comment by cefn** [comment]
I also experience this problem, arising from running in the wrong folder. 

On the plus side, https://github.com/kwai-explore/vscode-vitest-runner just works, although it doesn't have integration with the test explorer. The way the other one works correctly - first `cd` to the directory containing the target file. 

Running from the right folder means that tooling benefits from the test-specific tsconfig.json I have at `packages/packageName/test/tsconfig.json` in my monorepo and the vitest.config.ts I have at `packageName/vitest.config.ts` which references the tsconfig. Both are ignored if you run from the top level folder of the monorepo.

**Comment by cefn** [comment]
Worse, however, and really must be fixed, is the poor behaviour when the vitest process fails (owing to a poorly selected working directory). 

The UI simply swallows errors and sits there indefinitely with spinners rather than finalising the task with an error case.

![image](https://user-images.githubusercontent.com/159819/211013223-238038e4-c976-48f0-bb50-75bac1b092cf.png)

**Comment by cefn** [comment]
A workaround is to edit the vitest command to include an explicit root. For me, adding the following to the workspace meant that at least one of the packages in the monorepo could show interactive test results correctly.

```json
    "vitest.commandLine": "npx vitest --root packages/packageName/test"
```

In terms of the `vitest-dev` extension, if running from the proper root can't be configured as default behaviour, perhaps the basedir of the file being run could be passed as a variable, meaning users could inline this logic to the `vitest.commandLine` as required. 

This would lead to something like the following, which would mean more than one package in a monorepo could have passing tests at one time...

```json
    "vitest.commandLine": "npx vitest --root ${baseDir}"
```

**Comment by kasperstorgaard** [comment]
> Hi! I've had the same issue and I fixed it by creating [a VSCode Workspace file](https://code.visualstudio.com/docs/editor/workspaces). Once you have one the extension seems to pick up it needs to use the workspace root your test is in 👍

@adrienbaron I have tried to fix this through various ways, but cannot get vscode to auto detect the workspace and find the vite.config.ts file.
Can you share an example of you structure and vscode settings / workspace files?

**Comment by sean-sbl-uk** [comment]
> > > Hi! I've had the same issue and I fixed it by creating [a VSCode Workspace file](https://code.visualstudio.com/docs/editor/workspaces). Once you have one the extension seems to pick up it needs to use the workspace root your test is in 👍
> > 
> > 
> > You mean using [Multi-root workspaces](https://code.visualstudio.com/docs/editor/workspaces#_multiroot-workspaces)? Can you show me an example?
> 
> The project I've used that in is closed source (it's at my workplace), but it's literally just creating the file like indicated in the doc and setting one of the root path to be the folder where Vitest is installed in your repo 😉

Can you please provide a more detailed answer for those of us who are still struggling. Like kasperstorgaard above asked, maybe an example or screenshot would be very helpful.

**Comment by tomdavidson** [comment]
VS Code's multi-root workspace may work for some but for many of our shared/mono repo set ups, we have one VS Code root, one `.vscode/*` one `.git/*` etc. 

Rather than a multi-root issue this seems more of  `cwd` concern.

**Comment by tomdavidson** [comment]
Just messing around but here's a start: `"vitest.commandLine": "pnpm --if-present -r test --"`  pnpm has some create filtering options  to employ too.

**Comment by sfedorov-at-wiley** [comment]
Can confirm all of the above
We started using Nx and found out that extension does not recognise `/libs/ui/vite.config.ts` or `/apps/front/vite.config.ts`
There is no `vite.config.ts` in the root of repository
For every test in the repo extension printed "tests not found"
When we tried to define command in settings.json as `npx vitest --` all of the tests ended up in infinite loop

Hope this feedback helps

**Comment by ButuzGOL** [comment]
This workaround works for subfolders

```"vitest.commandLine": "npx vitest --root packages/packageName"```

Check logs in vitest in output it more informative

Only problem i am facing now it skips env variables

**Comment by mrcaidev** [comment]
This works for me, using Vitest Workspace:

```/vitest.workspace.ts
import { defineWorkspace } from "vitest/config";

export default defineWorkspace(["packages/*", "apps/*"]);
```

And nothing else.

Plus, I install vitest only in monorepo root.

**Comment by itsjavi** [comment]
> I believe vitest 0.30+ introduced the concept of [vitest workspaces](https://vitest.dev/guide/workspace.html)
> 
> @mrcaidev’s comment above is using the vitest.workspace.ts file as described in the vitest workspace docs.
> 
> After upgrading to vitest 0.30.1 and adding the workspace file to my root dir, the vscode extension started running tests in my packages/* folder correctly. 😄

can you please share your setup? I did that and it's still stuck.

I have vitest installed in the root package.json with pnpm,  and a vitest.workspace.ts:

```ts
import { defineWorkspace } from 'vitest/config'

export default defineWorkspace(['packages/*'])
```
what's your command for `vitest.commandLine` ?

**Comment by mrcaidev** [comment]
> what's your command for `vitest.commandLine` ?

I did not set `vitest.commandLine` myself so I guess the default would work.

**Comment by guzart** [comment]
> I did not set vitest.commandLine myself so I guess the default would work.

Same here. I added the workspaces config file in the root of the monorepo project:

```
// vitest.workspace.ts
export default ["packages/*"];
```

Then I made sure I was able to run `vitest` in watch mode successfully from one of the projects e.g. `packages/project-a/`. And finally I think the VSCode vitest extension had a conflict with another of my VSCode test related extensions, because it all started working after I disabled other VSCode test explorer extensions.

**Comment by jrson83** [comment]
Please provide a working example on how to setup vitest workspaces in a monorepo, with packages including other compilers then vite. After 5 hours I still can't get it to work.

**Comment by koistya** [comment]
I have a few repositories for experimenting with Vitest in a monorepo setup:

https://github.com/koistya/vitest-workspaces — bare minimum Vitest config
https://github.com/kriasoft/react-starter-kit — front-end monorepo
https://github.com/kriasoft/relay-starter-kit — full-stack monorepo

**Comment by daniel-johnsson** [comment]
I got it to work using the vs code workspace workaround. You can define your workspace using a file in the root with the name `yourmainfolder.code-workspace` and in that folder you can add a custom `vitest.commandLine` to each folder:

``` json
{
  "folders": [
    {
      "path": "apps/webapp",
      "vitest.commandLine": "yarn vitest --root apps/webapp"
    },
    {
      "path": "packages/integrations",
      "vitest.commandLine": "yarn vitest --root packages/integrations"
    }
  ]
}
```

**Comment by steven87vt** [comment]
In nx mono-repository I was able to get this working finally using the vitest.workspace.ts method, and nothing more then the default command line options:

Notes:

1. Use the project workspace settings to limit tests in the vitest ui. Jest tests should be excluded, or by exclusive include.
3. Use `vitest.workspace.ts` to isolate `vitest` enabled packages using the config file glob, seems to be needed for `vitest` to work at the command line.
4. Don't try to run `vitest --run` in mode via project settings, rather leave it in watch as the `vitest` plugin needs this to report success/failure cases.
5. I have come to avoid using the `import { configDefaults } from 'vitest/config'` package object values. I prefer a global config I can override in specific packages (e.g. `vitest.shared.ts` defines majority of test configurations)
6. Don't forget to absorb your, hopefully, already working `vite.config.ts` that should define `viteTsConfigPaths({ root: '../../../' })` plugin

```ts
// <root>/vitest.workspace.ts
import { defineWorkspace } from 'vitest/config'

export default defineWorkspace([
  'packages/**/vitest.config.{e2e,unit}.ts',
  'packages/**/vitest.config.ts'
])
```

```ts
// <root>/vitest.shared.ts

import { UserConfig } from 'vitest'
import { defaultExclude } from 'vitest/config'

const config: UserConfig = {
  globals: true,
  environment: 'node',
  cache: {
    dir: '../../../node_modules/.vitest'
  },
  include: ['test/**/*.test.ts'],
  exclude: defaultExclude
}

export default config
```

```ts
// <root>/packages/features/some-vue-feature/vitest.config.ts

import configShared from '../../../vitest.shared'
import { defineConfig, mergeConfig } from 'vitest/config'

import viteConfig from './vite.config'
import path from 'path'
import { UserConfig } from 'vitest'

const config: UserConfig = {
  ...configShared,
  environment: 'jsdom',
  setupFiles: ['test/unit/test-setup.ts'],
  alias: {
    '@/store': path.resolve(__dirname, 'test/unit/__mocks__/store.ts'),
    '@ezwebproductions/lib-core': path.resolve(__dirname, '../../lib/core/src/index.ts')
  }
}

export default mergeConfig(
  viteConfig({ mode: 'test', command: 'build' }),
  defineConfig({
    test: config
  })
)
```

```ts
// <root>/packages/lib/core/vitest.config.ts

import { mergeConfig, defineConfig } from 'vitest/config'
import configShared from '../../../vitest.shared'
import viteConfig from './vite.config'

export default mergeConfig(
  viteConfig({ command: 'build', mode: 'test' }),
  defineConfig({ test: configShared })
)
```

```jsonc
// project.code-workspace
{
    "vitest.enable": true,
    "vitest.commandLine": "npx vitest",
    "vitest.exclude": [
      "**/node_modules/**",
      "**/dist/**",
      "**/cypress/**",
      "**/.{idea,git,cache,output,temp}/**",
      "packages/commonjs-only-systems/**/*.test.ts"
    ],
    "vitest.include": [
      "packages/limit/your/use/case/for/ui/tests/here/**/*.test.ts",
      "packages/features/**/*.test.ts",
      "packages/lib/idb/**/*.test.ts",
      "packages/lib/core/**/*.test.ts"
    ]
}

**Comment by divmgl** [comment]
@steven87vt this was massively helpful. By far one of the most helpful comments I've seen on GitHub. Thanks a million for this!

**Comment by ffMathy** [comment]
This has now been properly fixed by merging #231 and #241!

Enjoy. Monorepo support should be a lot easier now, and a new NX example has been added as well. I think we can close this.
